src/entities/supportRequest.ts

```diff
@@ -0,0 +1,126 @@
+    this.type = type;
+  }
+
+  @PrimaryGeneratedColumn()
+  id: number;
+
+  @CreateDateColumn()
+  createdAt: Date;
+
+  @UpdateDateColumn()
+  updatedAt: Date;
+
+  @Column({ nullable: true })
+  movedToInProgressAt: Date;
+
+  @Column({ nullable: false })
+  slackId: string;
+
+  @Column({ nullable: false })
+  name: string;
+
+  @Column({ nullable: false, default: SupportRequestStatus.Pending, type: 'simple-enum', enum: SupportRequestStatus })
+  status: SupportRequestStatus;
+
+  @Column({ nullable: false, type: 'simple-enum', enum: SupportRequestType })
+  type: SupportRequestType;
+
+  @BeforeInsert()
+  async checkForCurrentRequest(): Promise<void> {
+    const abandonedRequests: SupportRequest[] = [];
+
+    const existingActiveRequests = await SupportRequest.find({
+      where: [{ slackId: this.slackId, status: SupportRequestStatus.Pending }, { slackId: this.slackId, status: SupportRequestStatus.InProgress }],
+    });
+
+    const minutesUntilStale = 15;
+
+    for (let i = 0; i < existingActiveRequests.length; i += 1) {
+      const { movedToInProgressAt, status } = existingActiveRequests[i];
+      const minutesInProgress = Math.abs(DateTime.fromJSDate(movedToInProgressAt || new Date()).diffNow('minutes').minutes);
+      if (status === SupportRequestStatus.InProgress && minutesInProgress >= minutesUntilStale) {
+        abandonedRequests.push(existingActiveRequests[i]);
+      }
+    }
+
+    if (existingActiveRequests.length && existingActiveRequests.length !== abandonedRequests.length) {
+      const error = new Error('Active support request exists for slackId');
+      error.name = SupportRequestErrors.ExistingActiveRequest;
+      throw error;
+    }
+
+    for (let i = 0; i < abandonedRequests.length; i += 1) {
+      try {
+        abandonedRequests[i].status = SupportRequestStatus.Abandoned;
+        await abandonedRequests[i].save();
+      } catch (err) {
+        // Ignore coverage for this line
+        /* istanbul ignore next */
+        logger.error('Something went wrong updating stale requests', err);
+      }
+    }
+  }
+
+  static async getNextSupportRequest(): Promise<SupportRequest> {
+    let retries = 5;
+
+    do {
+      const nextRequest = await SupportRequest.createQueryBuilder('supportRequests')
+        .orderBy('"supportRequests"."createdAt"', 'ASC')
+        .where({
+          status: SupportRequestStatus.Pending,
+        })
+        .getOne();
+
+      if (!nextRequest) {
+        return null;
+      }
+
+      try {
+        await SupportRequest.createQueryBuilder('supportRequests')
+          .update()
+          .where({
+            id: nextRequest.id,
+            status: nextRequest.status,
+          })
+          .set({
+            status: SupportRequestStatus.InProgress,
+            movedToInProgressAt: new Date(),
+          })
+          .execute();
+
+        await nextRequest.reload();
+        return nextRequest;
+      } finally {
+        retries -= 1;
+      }
```

Comment:
Not sure I understand this. How could this fail? Because you've already returned `null` if the `nextRequest` doesn't exist, and so reloading will never throw an error.

Reply:
Two requests exist: 1 and 2
User 1 and user 2 `getNext` at the same time and both get request 1, one of the two will successfully update the record, the other will not. The one who fails will retry and get 2. If there are no records, it just bails entirely.

Reply:
But what is actually failing? The update won't fail because it'll just update 0 records, and the reload won't fail because that has the record. Basically they would both get the same one.

Also, what happens when all the retries go through? It looks like it doesn't actually return anything if it has to use all retries